File: Beakjoon/Python/1697 숨바꼭질.py
from collections import deque

n, k = map(int, input().split())
graph = [[] for i in range(200001)]
visited = [-1 for i in range(200001)]

for i in range(100001) :
    if i != 0 : graph[i].append(i-1)
    graph[i].append(i+1)
    graph[i].append(i*2)
    
def bfs(start) :
    queue = deque()
    queue.append(start)
    visited[start] = 0
    while queue :
        v = queue.popleft()
        for to in graph[v] :
            if visited[to] == -1 :
                visited[to] = visited[v] + 1
                queue.append(to)

bfs(n)
print(visited[k])

File: Beakjoon/Cpp/1697 숨바꼭질.cpp
#include <bits/stdc++.h>

using namespace std;

vector<vector<int>> graph(200004, vector<int>());
bool visited[200004];
int dist[200004];
int n, k;

void bfs(int start) {
    
    dist[start] = 0;
    visited[start] = true;
    queue<int> queue;
    queue.push(start);
    
    while(!queue.empty()) {
        int v = queue.front(); queue.pop();
        for(int x : graph[v]) {
            if(!visited[x]) {
                visited[x] = true;
                dist[x] = dist[v] + 1;
                queue.push(x);
            }
        }
    }
    
    
    
}

int main() {
    cin >> n >> k;

    fill_n(dist, 200004, -1);
    for(int i = 0; i < 100002; i++) {
        
        if(i != 0) graph[i].push_back(i-1);
        graph[i].push_back(i+1);
        graph[i].push_back(i*2);
    }
    
    bfs(n);

    cout << dist[k];
}
